### ISSUE TITLE
Adding ASGI support?

### ISSUE THREAD

**Issue by arcivanov** [benoitc/gunicorn#1380]
[Django Daphne](https://github.com/django/daphne/issues/56) is under development to support ASGI, but Daphne server is extremely new and not yet suitable for production (e.g. no SSL support, maturation pains).

Is there any interest in adding [ASGI support](http://channels.readthedocs.io/en/latest/asgi.html) in GUnicorn to allow seamless WSGI/ASGI use?

**Comment by benoitc** [comment]
@arcivanov I will have a closer look at this protocol to see what needs to be done for it :)

Will update this ticket ASAP.

**Comment by japrogramer** [comment]
any updates?

**Comment by tilgovi** [comment]
@benoitc @berkerpeksag and I chatted over e-mail about it. No work has started, yet, that I know about. I have not had time yet to read more about ASGI, so I cannot estimate the complexity of implementing this. If anyone wants to contribute work toward this, I would enthusiastically help, discuss and review.

**Comment by andrewgodwin** [comment]
I'm in the process of overhauling the ASGI spec to make a lot more sense for servers anyway, so I'd caution about doing anything until that's finished (new spec is being drafted here, but is not final at all: http://channels.readthedocs.io/en/2.0/asgi.html - it's basically a lot more like WSGI)

**Comment by notedit** [comment]
@andrewgodwin   cool   can not wait to have a try gunicorn with asgi

**Comment by sirex** [comment]
@andrewgodwin what is the ASGI status?

I recently tried to test an ASGI app and here is what I got:

```python
import json

def app(scope):
    async def channel(receive, send):
        message = await receive()

        if scope['method'] == 'POST':
            response = message
        else:
            response = scope

        await send({
            'type': 'http.response.start',
            'status': 200,
            'headers': [
                [b'Content-Type', b'application/json'],
            ],
        })
        await send({
            'type': 'http.response.body',
            'body': json.dumps(response, default=bytes.decode).encode(),
        })
        await send({
            'type': 'http.disconnect',
        })
    return channel
```

```
> daphne app:app
2018-03-31 22:28:10,823 INFO     Starting server at tcp:port=8000:interface=127.0.0.1
2018-03-31 22:28:10,824 INFO     HTTP/2 support enabled
2018-03-31 22:28:10,824 INFO     Configuring endpoint tcp:port=8000:interface=127.0.0.1
2018-03-31 22:28:10,825 INFO     Listening on TCP address 127.0.0.1:8000
127.0.0.1:43436 - - [31/Mar/2018:22:28:17] "GET /" 200 347
127.0.0.1:43440 - - [31/Mar/2018:22:28:22] "POST /" 200 43
127.0.0.1:43446 - - [31/Mar/2018:22:28:42] "POST /" 200 54
```

```
> http -b get :8000/
{
    "type": "http"
    "http_version": "1.1",
    "method": "GET",
    "path": "/",
    "query_string": "",
    "root_path": "",
    "scheme": "http",
    "headers": [
        ["host", "localhost:8000"],
        ["user-agent", "HTTPie/0.9.9"],
        ["accept-encoding", "gzip, deflate"],
        ["accept", "*/*"],
        ["connection", "keep-alive"]
    ],
    "client": ["127.0.0.1", 43360],
    "server": ["127.0.0.1", 8000],
}

> http -b -f post :8000/ foo=bar
{
    "body": "foo=bar",
    "type": "http.request"
}

> http -b -j post :8000/ foo=bar
{
    "body": "{\"foo\": \"bar\"}",
    "type": "http.request"
}
```

https://github.com/django/asgiref/blob/master/specs/www.rst

**Comment by andrewgodwin** [comment]
@sirex ASGI is stable now. I'm not sure if you are saying that daphne does or does not adhere to the spec?

**Comment by benoitc** [comment]
is the link on django the last about the ASGI "spec", or is there anything more that describe the flow ?

**Comment by sirex** [comment]
@andrewgodwin I was trying to demostrate, that if Daphne supports ASGI, then ASGI must be ready for other frameworks to adopt.

@benoitc there are two ASGI specification pages:

https://github.com/django/asgiref/blob/master/specs/asgi.rst - general ASGI specification.

https://github.com/django/asgiref/blob/master/specs/www.rst - specifications describing HTTP and WebSockets protocols and compatibility with WSGI.

**Comment by sirex** [comment]
https://github.com/aio-libs/aiohttp/issues/2902

**Comment by benoitc** [comment]
@sirex thanks. Though these specs are not very helpful when it's about implementing. 

Anyway I have found https://channels.readthedocs.io/en/latest/ which is okish. I think i would prefer a pep, but we can propose something supporting asgi as soon as we removed the python 2 support. 

btw Ideleted last comment as it's not related to that discussion.

**Comment by erm** [comment]
The ASGI spec documentation is now available here: https://asgi.readthedocs.io/en/latest/.

**Comment by tomchristie** [comment]
Okay, have this covered now. (As a third-party worker class.)

Uvicorn 0.2 no longer depends on gunicorn for running it directly, but includes a two gunicorn worker classes for running ASGI apps.

For local dev, or cases where you don't care enough about process supervision you can just run uvicorn directly. For production we document using a gunicorn class as the preferred option.

http://www.uvicorn.org/#running-with-gunicorn

Run ASGI app from gunicorn, using uvloop and httptools:

```shell
gunicorn app:App -w 4 -k uvicorn.workers.UvicornWorker
```

Run ASGI app from gunicorn, using asyncio and h11 (for pypy compatibility):

```shell
gunicorn app:App -w 4 -k uvicorn.workers.UvicornH11Worker
```

If there any desire to bringing either of the protocol implementations into the core gunicorn package then am very open to considering that.

**Comment by waghanza** [comment]
Now that `gunicorn` does not support `python` `2` anymore,  it's technically doable to add `ASGI` support

**Comment by tilgovi** [comment]
@tomchristie I'm interested in what that would look like. As an alternative, maybe extracting the protocols as a separate package or packages?

**Comment by tomchristie** [comment]
> I'm interested in what that would look like. As an alternative, maybe extracting the protocols as a separate package or packages?

I guess either a worker class in gunicorn that depended on `uvicorn` for the h11 and/or httptools protocols. Or else duplicate the implementation here.

> maybe extracting the protocols as a separate package or packages?

There's not really much to extract from `uvicorn` - If we dropped `click` and just used argparse, and tweaked our setup.py dependencies a little then it'd have zero hard dependencies, with various options for different HTTP implementations, WS implementations, Event loops.

**Comment by tilgovi** [comment]
I'd like first-class ASGI support, whether that means adding extras dependencies in Gunicorn, a stub worker that tells you what to install, documentation, or something else.

**Comment by japrogramer** [comment]
yes, please. about to switch to daphne from gunicorn .. if gunicorn could just support asgi
that would save me so much time.

**Comment by tilgovi** [comment]
@japrogramer http://www.uvicorn.org/#running-with-gunicorn

**Comment by johnthagen** [comment]
[Django 3.0 (set to release December 2019) will have ASGI support out of the box](https://docs.djangoproject.com/en/dev/releases/3.0/#asgi-support), so that will be another great reason for `gunicorn` to support it directly. We'll probably see a big up-tick in user interest after that is released.

**Comment by benoitc** [comment]
@johnthagen that's on the pipe. However, we need to clear up a little the tickets and make a release this month before starting .

**Comment by mirekphd** [comment]
@benoitc any progress on this issue?

**Comment by Ahleroy** [comment]
Also interested in the ASGI support so we do not need to use both GUnicorn and UVicorn.

**Comment by AmeerArsala** [comment]
AmeerArsala

**Comment by acuencadev** [comment]
Any news here?

**Comment by petrprikryl** [comment]
> Anybody alive?

Maybe you can try https://github.com/emmett-framework/granian

**Comment by blopker** [comment]
Wow, granian looks amazing, thanks for sharing.

**Comment by benoitc** [comment]
@alexted no. thanks for your contributions to this ticket.

**Comment by johnthagen** [comment]
@alexted Uvicorn supports ASGI natively and has it's own built in process manager (an alternative to Gunicorn). You could try

- https://www.uvicorn.org/deployment/#built-in

**Comment by benoitc** [comment]
please keep tickets focused on Gunicorn. I's OK to discuss about this issue or  compatible worker but not about anything else. @alexted work is in progress. Anf yes we are alive. A new version came out just this weekend. Maybe you wantrf to ask about the progress of this issue? 

Please look again latest discussions but work is in progress to suport the python async API.

**Comment by benoitc** [comment]
please keep tickets focused on Gunicorn. I's OK to discuss about this issue or  compatible worker but not about anything else. 

@alexted project if defin tly alive. But maybe you wanted to ask about the status of this ticket? Work is in progress to support the async semantic of python. ASGI will be supported through it. This will land soon.

**Comment by breid1313** [comment]
@benoitc do we have a rough estimate on when this might land?

**Comment by xdevfaheem** [comment]
Even after a ~decade, the support is still yet to arrive. Yikes !

**Comment by alexted** [comment]
Forget about gunicorn - this is how open source projects die. You're seeing it right now.

**Comment by malthejorgensen** [comment]
Alex and Mohammed, don't be such downers. I've used gunicorn in production for more than a decade, works very well. Still works very well even with the uvicorn async workers. Benoit is lifting the Python web world up on his shoulders mostly _alone_ as far as I can see. 

Would love to donate.

Thank you for all your hard work Benoit <3

**Comment by xdevfaheem** [comment]
Fair enough, gunivorn + unicorn + http-tools + uvloop... has been pretty solid.

But Granian seems pretty cool btw, I've been eyeing that a lot lately...

h/t @gi0baro

**Comment by mtvx** [comment]
@benoitc I see you've recently added ASGI worker, with the [first commits in Jan 2026](https://github.com/benoitc/gunicorn/commits/master/gunicorn/asgi).

The sources are in https://github.com/benoitc/gunicorn/tree/master/gunicorn/asgi.

The [25.1.0 release](https://gunicorn.org/2026-news/#2510-2026-02-13) (2026-02-13) mentions:

> ASGI Worker: Promoted from beta to stable

How stable are we, is it already something you'd use in production? (Based on "stable" flag, I'd assume yes.)

Do you have any performance figures, especially against likes of Uvicorn or Granian?

For our use case (the typical JSON HTTP API with DB and stuff) the speed of the ASGI runtime doesn't really matter... but I assume you have some hunch at least?

---

Update: The ASGI worker doesn't seem to support `worker-connections` argument (despite [the docs suggesting so](https://gunicorn.org/asgi/#recommended-settings))? I'd like to restrict how many concurrent requests each worker accepts. Do you plan to support it?

**Comment by alexted** [comment]
Whoa, holy cow! Gunicorn finally dropped native async support?! I'm literally shaking, can't believe it's happening!
Huge congrats to those of us who stayed alive and sane through the wait, to everyone still holding out, and to the devs! "That’s one small step for man, one giant leap for mankind."

**Comment by fseasy** [comment]
Hey, so for asgi (or specifically, fastapi), what should i do:
1. as gunicorn doc: `--worker-class asgi`
2. or as fastapi doc: `-k uvicorn.workers.UvicornWorker`

Which one is better?

**Comment by johnthagen** [comment]
Would be great to have some independent benchmarking comparing unicorn to this new worker.

**Comment by benoitc** [comment]
@tuukkamustonen There is still room for performance improvements, and I am currently working on it. I have a release due this week, so I will focus more on this topic shortly after.
I temporarily removed the limit during the release process, and a PR is already available. I will review it and likely include it in an upcoming release.
@fseasy I am currently using ASGI myself.

**Comment by alexted** [comment]
I don't know why, but another library for this has appeared: https://github.com/Kludex/uvicorn-worker

**Comment by benoitc** [comment]
I'm closing this issue since ASGI is supported already. For perfmances improvements yopu may want to track https://github.com/benoitc/gunicorn/pull/3549
